File: bubble_sort.rb
class BubbleSort

  def sort(data)

    sorted_data = []
    
    until data.length == 0 do

      previous = data[0]
      current = data[1]
      pass_complete = false
      index = 1

      until pass_complete == true do

        if data.length > 1 && previous > current
            temporary_holder = previous
            data[index - 1] = current
            data[index] = temporary_holder
        end

            index += 1

            previous = data[index - 1]
            current = data[index]

        if index == (data.length) || data.length == 1
          sorted_data.unshift(data.pop)
          pass_complete = true
        end

      end

    end 

    return sorted_data 
    
  end

end

# Make some data and an instance of BubbleSort. Call the sort method and print the results to the terminal.
# data = [0, 3, 1, 2, 5, 100, 4, 3]
data = ["d", "b", "a", "c"]
bubble_sorter = BubbleSort.new
puts bubble_sorter.sort(data)

File: bubble_sort/js/bubble_sort.js
window.onload = function bubbleSort() {

  var data = [];
  for (i = 0; i < 7; i++){
    data[i] = Math.floor(Math.random() * 100);
  }

  // Put the unsorted data in the DOM
  for (i = 0; i < data.length; i++) {

    var elementContainer = document.createElement("div");
    elementContainer.className = "element-container";

    var element = document.createElement("div");
    element.innerHTML = data[i];

    elementContainer.appendChild(element);
    document.getElementById("unsorted-elements-container").appendChild(elementContainer);

  }  

  var sorted_data = [];

  // Sort the data
  while (data.length > 0) {

    var previous = data[0];
    var current = data[1];
    var pass_complete = false;
    var index = 1;

    while (pass_complete == false) {

      if (data.length > 1 && previous > current) {
        var temporary_holder = previous;
        data[index - 1] = current;
        data[index] = temporary_holder;
      }

      index ++;

      previous = data[index - 1];
      current = data[index];

      if (index == data.length || data.length == 1) {
        sorted_element = data.pop();
        sorted_data.unshift(sorted_element);
        pass_complete = true;
      }
    }

  }

  // Put the sorted data in the DOM
  for (i = 0; i < sorted_data.length; i++) {

    var elementContainer = document.createElement("div");
    elementContainer.className = "element-container";

    var element = document.createElement("div");
    element.innerHTML = sorted_data[i];

    elementContainer.appendChild(element);
    document.getElementById("sorted-elements-container").appendChild(elementContainer);

  }  

}

function randomizeData() {
  location.reload();
}
